src/main/java/org/zalando/nakadi/service/subscription/zk/CuratorZkSubscriptionClient.java

```diff
@@ -317,4 +326,101 @@ public ZkSubscriptionNode getZkSubscriptionNodeLocked() {
+        };
+    }
+
+    @Override
+    public boolean isCursorResetInProgress() {
+        try {
+            return curatorFramework.checkExists().forPath(resetCursorPath) != null;
+        } catch (final Exception e) {
+            // nothing in the path
+        }
+        return false;
+    }
+
+    @Override
+    public void resetCursors(final List<NakadiCursor> cursors, final Long timeout) throws OperationTimeoutException,
+            ZookeeperException, OperationInterruptedException, RequestInProgressException{
+        ZKSubscription sessionsListener = null;
+        boolean resetWasAlreadyInitiated = false;
+        try {
+            // close subscription connections
+            curatorFramework.create().withMode(CreateMode.EPHEMERAL).forPath(resetCursorPath);
+
+            final AtomicBoolean sessionsChanged = new AtomicBoolean(true);
+            sessionsListener = subscribeForSessionListChanges(() -> sessionsChanged.set(true));
+
+            final String sessionsPath = MessageFormat.format("/nakadi/subscriptions/{0}/sessions", subscriptionId);
+            final long finishAt = System.currentTimeMillis() + timeout;
+            while (finishAt > System.currentTimeMillis()) {
+                if (sessionsChanged.compareAndSet(true, false) &&
+                        curatorFramework.getChildren().forPath(sessionsPath).isEmpty()) {
+                    for (final NakadiCursor cursor : cursors) {
+                        final String path = MessageFormat
+                                .format("/nakadi/subscriptions/{0}/topics/{1}/{2}/offset",
+                                        subscriptionId, cursor.getTopic(), cursor.getPartition());
+                        curatorFramework.setData().forPath(path, cursor.getOffset().getBytes(Charsets.UTF_8));
+                    }
+                    return;
+                }
+
+                Thread.sleep(100);
+            }
+        } catch (final InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new OperationInterruptedException("Resetting cursors is interrupted", e);
+        } catch (final KeeperException.NodeExistsException e) {
+            resetWasAlreadyInitiated = true;
+            throw new RequestInProgressException("Cursors reset is already in progress for provided subscription", e);
+        } catch (final Exception e) {
+            throw new ZookeeperException("Unexpected problem occurred when resetting cursors", e);
+        } finally {
+            if (sessionsListener != null) {
+                sessionsListener.cancel();
+            }
+
+            try {
+                if (!resetWasAlreadyInitiated) {
+                    curatorFramework.delete().forPath(resetCursorPath);
+                }
+            } catch (final KeeperException.NoNodeException e) {
+                throw new UnableProcessException(
+                        "Impossible to reset cursors. Subscription has not been streamed yet", e);
```

Comment:
Well, for me it seems that reset was correctly processed(or real reason of problem is already thrown), we just failed to clean up. I think that we may log this, but it is definitely not the reason to say that reset failed.